src/eos-updater-poll-main.c

```diff
@@ -0,0 +1,224 @@
+  query = soup_form_encode_hash (hw_descriptors);
+  if (!try_all_branch_file_sources (repo,
+                                    baseurl,
+                                    query,
+                                    &branch_file_contents,
+                                    &signature_contents,
+                                    error))
+    return FALSE;
+
+  now = g_date_time_new_now_utc ();
+  branch_file = eos_branch_file_new_from_raw (branch_file_contents,
+                                              signature_contents,
+                                              now,
+                                              error);
+  if (branch_file == NULL)
+    return FALSE;
+
+  *out_branch_file = g_steal_pointer (&branch_file);
+  return TRUE;
+}
+
+gboolean
+metadata_fetch_from_main (EosMetadataFetchData *fetch_data,
+                          EosUpdateInfo **info,
+                          EosMetricsInfo **out_metrics,
+                          GError **error)
+{
+  OstreeRepo *repo = fetch_data->data->repo;
+  g_autofree gchar *refspec = NULL;
+  g_autofree gchar *orig_refspec = NULL;
+  g_autoptr(EosBranchFile) branch_file = NULL;
+  g_autoptr(EosMetricsInfo) metrics = NULL;
+
+  g_return_val_if_fail (info != NULL, FALSE);
+
+  if (!download_branch_file (repo,
+                             &branch_file,
+                             error))
+    return FALSE;
+
+  if (!get_upgrade_info_from_branch_file (branch_file,
+                                          &refspec,
+                                          &orig_refspec,
+                                          &metrics,
+                                          error))
+    return FALSE;
+
+  if (!metrics->on_hold)
+    {
+      g_autofree gchar *checksum = NULL;
+      g_autoptr(GVariant) commit = NULL;
+      g_autofree gchar *remote = NULL;
+      g_autofree gchar *ref = NULL;
+      g_autoptr(EosExtensions) extensions = NULL;
+
+      if (!ostree_parse_refspec (refspec, &remote, &ref, error))
+        return FALSE;
+
+      if (!fetch_latest_commit (repo,
+                                g_task_get_cancellable (fetch_data->task),
+                                remote,
+                                ref,
+                                NULL,
+                                &checksum,
+                                &extensions,
+                                error))
+        return FALSE;
+
+      if (!is_checksum_an_update (repo, checksum, &commit, error))
+        return FALSE;
+
+      g_set_object (&extensions->branch_file, branch_file);
+      if (commit != NULL)
+        *info = eos_update_info_new (checksum,
+                                     commit,
+                                     refspec,
+                                     orig_refspec,
+                                     NULL,
+                                     extensions);
```

Comment:
Right, we should set the *info to `NULL` explicitly, instead of relying on the caller to set it to NULL before calling this function. Fixed.